slither/utils/upgradeability.py

```diff
@@ -0,0 +1,463 @@
+                if isinstance(dest, Identifier):
+                    delegate = dest.value
+                    break
+                if isinstance(dest, Literal) and len(dest.value) == 66:
+                    delegate = StateVariable()
+                    delegate.is_constant = True
+                    delegate.expression = dest
+                    delegate.name = dest.value
+                    delegate.type = ElementaryType("bytes32")
+                    break
+    return delegate
+
+
+def extract_delegate_from_asm(contract: Contract, node: Node) -> Optional[Variable]:
+    """
+    Finds a Variable with a name matching the argument passed into a delegatecall, when all we have is an Assembly node
+    with a block of code as one long string. Usually only the case for solc versions < 0.6.0.
+    Should typically be called by find_delegate_in_fallback(proxy).
+    Args:
+        contract: The parent Contract.
+        node: The Assembly Node (i.e., node.type == NodeType.ASSEMBLY)
+
+    Returns:
+        (`Variable`) | None : The variable being passed as the destination argument in a delegatecall in the fallback.
+    """
+    asm_split = str(node.inline_asm).split("\n")
+    asm = next(line for line in asm_split if "delegatecall" in line)
+    params = asm.split("call(")[1].split(", ")
+    dest = params[1]
+    if dest.endswith(")"):
+        dest = params[2]
+    if dest.startswith("sload("):
+        dest = dest.replace(")", "(").split("(")[1]
+        for v in node.function.variables_read_or_written:
+            if v.name == dest:
+                if isinstance(v, LocalVariable) and v.expression is not None:
+                    e = v.expression
+                    if isinstance(e, Identifier) and isinstance(e.value, StateVariable):
+                        v = e.value
+                        # Fall through, return constant storage slot
+                if isinstance(v, StateVariable) and v.is_constant:
+                    return v
+    if "_fallback_asm" in dest or "_slot" in dest:
+        dest = dest.split("_")[0]
+    return find_delegate_from_name(contract, dest, node.function)
+
+
+def find_delegate_from_name(
+    contract: Contract, dest: str, parent_func: Function
+) -> Optional[Variable]:
+    """
+    Searches for a variable with a given name, starting with StateVariables declared in the contract, followed by
+    LocalVariables in the parent function, either declared in the function body or as parameters in the signature.
+    Args:
+        contract: The Contract object to search.
+        dest: The variable name to search for.
+        parent_func: The Function object to search.
+
+    Returns:
+        (`Variable`) | None : The variable with the matching name, if found
+    """
+    for sv in contract.state_variables:
+        if sv.name == dest:
+            return sv
+    for lv in parent_func.local_variables:
+        if lv.name == dest:
+            return lv
+    for pv in parent_func.parameters:
```

Comment:
maybe `parent_func.returns` too? In case the variable is declared in the return statement